File: 21.merge-two-sorted-lists.rb
#
# @lc app=leetcode id=21 lang=ruby
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}
def merge_two_lists(list1, list2)
  return list1 unless list2
  return list2 unless list1

  if list1.val > list2.val
    list2.next = merge_two_lists(list1, list2.next)
    list2
  else
    list1.next = merge_two_lists(list1.next, list2)
    list1
  end
end
# @lc code=end

File: 21.merge-two-sorted-lists.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 */

package main

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	root := &ListNode{}
	current := root

	for {
		if list1 == nil && list2 == nil {
			break
		}

		if list1 == nil {
			current.Next = list2
			list2 = list2.Next
		} else if list2 == nil {
			current.Next = list1
			list1 = list1.Next
		} else if list1.Val < list2.Val {
			current.Next = list1
			list1 = list1.Next
		} else {
			current.Next = list2
			list2 = list2.Next
		}

		current = current.Next
	}

	return root.Next

}

// @lc code=end

type ListNode struct {
	Val  int
	Next *ListNode
}
